File: scriptable/src/lightness-remap.js
/**
 * lightness-remap.js — CLAHE 明度リマッピング
 *
 * L* チャンネルに Contrast Limited Adaptive Histogram Equalization を適用し、
 * パレットの明度範囲を有効活用する。
 * Python実装 (infrastructure/lightness_remap.py) の忠実なJSポート。
 */

/**
 * 単チャンネルに対する CLAHE 実装。
 *
 * @param {Float64Array|number[]} channel - (H*W) フラット配列
 * @param {number} width - 画像幅
 * @param {number} height - 画像高さ
 * @param {number} clipLimit - コントラスト制限係数 (1.0=弱い, 4.0=強い)
 * @param {number} gridSize - グリッド分割数
 * @param {number} minVal - チャンネルの最小値
 * @param {number} maxVal - チャンネルの最大値
 * @param {number} [nBins=256] - ヒストグラムのビン数
 * @returns {Float64Array} CLAHE 適用後のフラット配列
 */
function claheChannel(channel, width, height, clipLimit, gridSize, minVal, maxVal, nBins) {
  nBins = nBins || 256;

  const result = new Float64Array(width * height);
  const valRange = maxVal - minVal;
  if (valRange < 1e-10) {
    result.set(channel);
    return result;
  }

  // 値を [0, nBins-1] にスケーリング
  const scaled = new Float64Array(width * height);
  const scaleF = (nBins - 1) / valRange;
  for (let i = 0; i < scaled.length; i++) {
    let v = (channel[i] - minVal) * scaleF;
    if (v < 0) v = 0;
    if (v > nBins - 1) v = nBins - 1;
    scaled[i] = v;
  }

  // グリッド境界
  const rowStep = height / gridSize;
  const colStep = width / gridSize;

  // 各グリッドブロックの CDF を事前計算
  // cdfs[gy][gx] = Float64Array(nBins)
  const cdfs = [];
  for (let gy = 0; gy < gridSize; gy++) {
    cdfs[gy] = [];
    const y0 = Math.round(gy * rowStep);
    let y1 = Math.round((gy + 1) * rowStep);
    if (y1 <= y0) y1 = y0 + 1;

    for (let gx = 0; gx < gridSize; gx++) {
      const x0 = Math.round(gx * colStep);
      let x1 = Math.round((gx + 1) * colStep);
      if (x1 <= x0) x1 = x0 + 1;

      const nPixels = (y1 - y0) * (x1 - x0);

      // ヒストグラム
      const hist = new Float64Array(nBins);
      for (let y = y0; y < y1; y++) {
        for (let x = x0; x < x1; x++) {
          let idx = Math.round(scaled[y * width + x]);
          if (idx < 0) idx = 0;
          if (idx >= nBins) idx = nBins - 1;
          hist[idx] += 1.0;
        }
      }

      // クリッピング
      const actualClip = clipLimit * nPixels / nBins;
      let excess = 0.0;
      for (let i = 0; i < nBins; i++) {
        if (hist[i] > actualClip) {
          excess += hist[i] - actualClip;
          hist[i] = actualClip;
        }
      }

      // 超過分を均等再分配
      const redistrib = excess / nBins;
      for (let i = 0; i < nBins; i++) {
        hist[i] += redistrib;
      }

      // CDF
      const cdf = new Float64Array(nBins);
      cdf[0] = hist[0];
      for (let i = 1; i < nBins; i++) {
        cdf[i] = cdf[i - 1] + hist[i];
      }

      let cdfMin = 0;
      for (let i = 0; i < nBins; i++) {
        if (cdf[i] > 0) { cdfMin = cdf[i]; break; }
      }

      const denom = nPixels - cdfMin;
      if (denom < 1.0) {
        for (let i = 0; i < nBins; i++) {
          cdf[i] = i;
        }
      } else {
        for (let i = 0; i < nBins; i++) {
          cdf[i] = (cdf[i] - cdfMin) / denom * (nBins - 1);
        }
      }

      cdfs[gy][gx] = cdf;
    }
  }

  // バイリニア補間で全ピクセルをリマッピング
  for (let y = 0; y < height; y++) {
    const gyF = (y + 0.5) / rowStep - 0.5;
    let gy0 = Math.floor(gyF);
    let gy1 = gy0 + 1;
    const fy = gyF - gy0;
    if (gy0 < 0) gy0 = 0;
    if (gy0 >= gridSize) gy0 = gridSize - 1;
    if (gy1 < 0) gy1 = 0;
    if (gy1 >= gridSize) gy1 = gridSize - 1;

    for (let x = 0; x < width; x++) {
      const gxF = (x + 0.5) / colStep - 0.5;
      let gx0 = Math.floor(gxF);
      let gx1 = gx0 + 1;
      const fx = gxF - gx0;
      if (gx0 < 0) gx0 = 0;
      if (gx0 >= gridSize) gx0 = gridSize - 1;
      if (gx1 < 0) gx1 = 0;
      if (gx1 >= gridSize) gx1 = gridSize - 1;

      const val = scaled[y * width + x];
      let idx = Math.floor(val);
      if (idx < 0) idx = 0;
      if (idx > nBins - 2) idx = nBins - 2;
      const frac = val - idx;

      // 4ブロックの CDF を線形補間
      const v00 = cdfs[gy0][gx0][idx] * (1 - frac) + cdfs[gy0][gx0][idx + 1] * frac;
      const v01 = cdfs[gy0][gx1][idx] * (1 - frac) + cdfs[gy0][gx1][idx + 1] * frac;
      const v10 = cdfs[gy1][gx0][idx] * (1 - frac) + cdfs[gy1][gx0][idx + 1] * frac;
      const v11 = cdfs[gy1][gx1][idx] * (1 - frac) + cdfs[gy1][gx1][idx + 1] * frac;

      // バイリニア補間
      const top = v00 * (1 - fx) + v01 * fx;
      const bot = v10 * (1 - fx) + v11 * fx;
      const mapped = top * (1 - fy) + bot * fy;

      result[y * width + x] = mapped / (nBins - 1) * valRange + minVal;
    }
  }

  return result;
}

/**
 * L* チャンネルに CLAHE を適用。
 * imageData を in-place で書き換える。
 *
 * @param {{data: Uint8ClampedArray, width: number, height: number}} imageData
 * @param {number} clipLimit - コントラスト制限 (1.0=弱い, 4.0=強い)
 * @param {number} [gridSize=8] - CLAHE グリッドサイズ
 */
function claheLightness(imageData, clipLimit, gridSize) {
  gridSize = gridSize || 8;
  const { data, width, height } = imageData;
  const n = width * height;

  // RGB → Lab (L* チャンネル抽出)
  const lChannel = new Float64Array(n);
  const aChannel = new Float64Array(n);
  const bChannel = new Float64Array(n);

  for (let i = 0; i < n; i++) {
    const lab = rgbToLab(data[i * 4], data[i * 4 + 1], data[i * 4 + 2]);
    lChannel[i] = lab[0];
    aChannel[i] = lab[1];
    bChannel[i] = lab[2];
  }

  // L* に CLAHE 適用
  const lEnhanced = claheChannel(lChannel, width, height, clipLimit, gridSize, 0.0, 100.0);

  // L* をクランプ
  for (let i = 0; i < n; i++) {
    let l = lEnhanced[i];
    if (l < 0.0) l = 0.0;
    if (l > 100.0) l = 100.0;
    lEnhanced[i] = l;
  }

  // Lab → RGB (in-place)
  for (let i = 0; i < n; i++) {
    const rgb = labToRgb(lEnhanced[i], aChannel[i], bChannel[i]);
    data[i * 4] = rgb[0];
    data[i * 4 + 1] = rgb[1];
    data[i * 4 + 2] = rgb[2];
  }
}

// Export for test and module usage
if (typeof module !== 'undefined' && module.exports) {
  module.exports = { claheChannel, claheLightness };
}

File: src/epaper_palette_dither/infrastructure/lightness_remap.py
"""明度リマッピング (CLAHE)。

L* チャンネルに Contrast Limited Adaptive Histogram Equalization を適用し、
パレットの明度範囲を有効活用する。
"""

from __future__ import annotations

import numpy as np
import numpy.typing as npt

from epaper_palette_dither.infrastructure.color_space import lab_to_rgb_batch, rgb_to_lab_batch


def _clahe_channel(
    channel: npt.NDArray[np.float64],
    clip_limit: float,
    grid_size: int,
    value_min: float,
    value_max: float,
    n_bins: int = 256,
) -> npt.NDArray[np.float64]:
    """単チャンネルに対する CLAHE 実装。

    Args:
        channel: (H, W) float64
        clip_limit: コントラスト制限係数 (1.0=弱い, 4.0=強い)
        grid_size: グリッド分割数
        value_min: チャンネルの最小値
        value_max: チャンネルの最大値
        n_bins: ヒストグラムのビン数

    Returns:
        CLAHE 適用後の (H, W) float64
    """
    h, w = channel.shape
    result = np.empty_like(channel)

    # 値を [0, n_bins-1] にスケーリング
    val_range = value_max - value_min
    if val_range < 1e-10:
        return channel.copy()
    scaled = (channel - value_min) / val_range * (n_bins - 1)
    scaled = np.clip(scaled, 0, n_bins - 1)

    # グリッド境界
    row_step = h / grid_size
    col_step = w / grid_size

    # 各グリッドブロックの CDF を事前計算
    cdfs = np.empty((grid_size, grid_size, n_bins), dtype=np.float64)

    for gy in range(grid_size):
        y0 = int(round(gy * row_step))
        y1 = int(round((gy + 1) * row_step))
        y1 = max(y1, y0 + 1)
        for gx in range(grid_size):
            x0 = int(round(gx * col_step))
            x1 = int(round((gx + 1) * col_step))
            x1 = max(x1, x0 + 1)

            block = scaled[y0:y1, x0:x1]
            n_pixels = block.size

            # ヒストグラム
            hist = np.zeros(n_bins, dtype=np.float64)
            indices = np.clip(block.astype(np.int32), 0, n_bins - 1)
            for val in indices.ravel():
                hist[val] += 1.0

            # クリッピング
            actual_clip = clip_limit * n_pixels / n_bins
            excess = 0.0
            for i in range(n_bins):
                if hist[i] > actual_clip:
                    excess += hist[i] - actual_clip
                    hist[i] = actual_clip

            # 超過分を均等再分配
            redistrib = excess / n_bins
            hist += redistrib

            # CDF
            cdf = np.cumsum(hist)
            cdf_min = cdf[cdf > 0].min() if np.any(cdf > 0) else 0.0
            denom = n_pixels - cdf_min
            if denom < 1.0:
                cdfs[gy, gx, :] = np.arange(n_bins, dtype=np.float64)
            else:
                cdfs[gy, gx, :] = (cdf - cdf_min) / denom * (n_bins - 1)

    # バイリニア補間で全ピクセルをリマッピング
    for y in range(h):
        # グリッド中心からの相対位置
        gy_f = (y + 0.5) / row_step - 0.5
        gy0 = int(np.floor(gy_f))
        gy1 = gy0 + 1
        fy = gy_f - gy0
        gy0 = max(0, min(grid_size - 1, gy0))
        gy1 = max(0, min(grid_size - 1, gy1))

        for x in range(w):
            gx_f = (x + 0.5) / col_step - 0.5
            gx0 = int(np.floor(gx_f))
            gx1 = gx0 + 1
            fx = gx_f - gx0
            gx0 = max(0, min(grid_size - 1, gx0))
            gx1 = max(0, min(grid_size - 1, gx1))

            val = scaled[y, x]
            idx = int(np.clip(val, 0, n_bins - 2))
            frac = val - idx

            # 4ブロックの CDF を線形補間
            v00 = cdfs[gy0, gx0, idx] * (1 - frac) + cdfs[gy0, gx0, idx + 1] * frac
            v01 = cdfs[gy0, gx1, idx] * (1 - frac) + cdfs[gy0, gx1, idx + 1] * frac
            v10 = cdfs[gy1, gx0, idx] * (1 - frac) + cdfs[gy1, gx0, idx + 1] * frac
            v11 = cdfs[gy1, gx1, idx] * (1 - frac) + cdfs[gy1, gx1, idx + 1] * frac

            # バイリニア補間
            top = v00 * (1 - fx) + v01 * fx
            bot = v10 * (1 - fx) + v11 * fx
            mapped = top * (1 - fy) + bot * fy

            result[y, x] = mapped / (n_bins - 1) * val_range + value_min

    return result


def clahe_lightness(
    rgb_array: npt.NDArray[np.uint8],
    clip_limit: float = 2.0,
    grid_size: int = 8,
) -> npt.NDArray[np.uint8]:
    """L* チャンネルに CLAHE を適用。

    Args:
        rgb_array: (H, W, 3) uint8 RGB
        clip_limit: コントラスト制限 (1.0=弱い, 4.0=強い)
        grid_size: CLAHE グリッドサイズ

    Returns:
        CLAHE 適用後の (H, W, 3) uint8 RGB
    """
    lab = rgb_to_lab_batch(rgb_array)
    l_star = lab[:, :, 0]

    l_enhanced = _clahe_channel(l_star, clip_limit, grid_size, 0.0, 100.0)
    l_enhanced = np.clip(l_enhanced, 0.0, 100.0)

    lab_enhanced = lab.copy()
    lab_enhanced[:, :, 0] = l_enhanced

    return lab_to_rgb_batch(lab_enhanced)
